File: app/src/main/java/com/micewine/emu/LorieView.java
package com.micewine.emu;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.ContextWrapper;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.graphics.PixelFormat;
import android.graphics.Point;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Color; 
import android.graphics.Rect;
import android.graphics.drawable.ColorDrawable;
import android.os.Handler;
import android.preference.PreferenceManager;
import android.util.AttributeSet;
import android.util.Log;
import android.view.KeyEvent;
import android.view.Surface;
import android.view.SurfaceHolder;
import android.view.SurfaceView;
import android.view.View;

import androidx.annotation.NonNull;

import com.micewine.emu.input.InputStub;

import com.micewine.emu.systemutils.SystemMemoryInfo;
import java.util.regex.PatternSyntaxException;

@SuppressLint("WrongConstant")
@SuppressWarnings("deprecation")
public class LorieView extends SurfaceView implements InputStub {

    private long totalMemory = SystemMemoryInfo.getTotalRAM(getContext());
    private long freeMemory = SystemMemoryInfo.getFreeRAM(getContext());
    private Handler handler = new Handler();
    public static long BYTES_FOR_MEGABYTES = (1024*1024);
    
    
    interface Callback {
        void changed(Surface sfc, int surfaceWidth, int surfaceHeight, int screenWidth, int screenHeight);
    }

    interface PixelFormat {
        int BGRA_8888 = 5; // Stands for HAL_PIXEL_FORMAT_BGRA_8888
    }

    private Callback mCallback;
    private final Point p = new Point();
    private final SurfaceHolder.Callback mSurfaceCallback = new SurfaceHolder.Callback() {
        @Override public void surfaceCreated(@NonNull SurfaceHolder holder) {
           updateRamCounter();
            holder.setFormat(PixelFormat.BGRA_8888);
        }

        @Override public void surfaceChanged(@NonNull SurfaceHolder holder, int f, int width, int height) {
            width = getMeasuredWidth();
            height = getMeasuredHeight();

            Log.d("SurfaceChangedListener", "Surface was changed: " + width + "x" + height);
            if (mCallback == null)
                return;

            
            
            getDimensionsFromSettings();
            mCallback.changed(holder.getSurface(), width, height, p.x, p.y);
        }

        @Override public void surfaceDestroyed(@NonNull SurfaceHolder holder) {
            if (mCallback != null)
                mCallback.changed(holder.getSurface(), 0, 0, 0, 0);
        }
    };

    public LorieView(Context context) { super(context); init(); }
    public LorieView(Context context, AttributeSet attrs) { super(context, attrs); init(); }
    public LorieView(Context context, AttributeSet attrs, int defStyleAttr) { super(context, attrs, defStyleAttr); init(); }
    @SuppressWarnings("unused")
    public LorieView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) { super(context, attrs, defStyleAttr, defStyleRes); init(); }

    private void init() {
        getHolder().addCallback(mSurfaceCallback);
    }

    public void setCallback(Callback callback) {
        mCallback = callback;
        triggerCallback();
    }

    public void regenerate() {
        Callback callback = mCallback;
        mCallback = null;
        getHolder().setFormat(android.graphics.PixelFormat.RGBA_8888);
        mCallback = callback;

        triggerCallback();
    }

    public void triggerCallback() {
        setFocusable(true);
        setFocusableInTouchMode(true);
        requestFocus();

        setBackground(new ColorDrawable(Color.TRANSPARENT) {
            public boolean isStateful() {
                return true;
            }
            public boolean hasFocusStateSpecified() {
                return true;
            }
        });

        Rect r = getHolder().getSurfaceFrame();
        getActivity().runOnUiThread(() -> mSurfaceCallback.surfaceChanged(getHolder(), PixelFormat.BGRA_8888, r.width(), r.height()));
    }

    private Activity getActivity() {
        Context context = getContext();
        while (context instanceof ContextWrapper) {
            if (context instanceof Activity) {
                return (Activity) context;
            }
            context = ((ContextWrapper) context).getBaseContext();
        }

        throw new NullPointerException();
    }

    void getDimensionsFromSettings() {
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(getContext());
        int width = getMeasuredWidth();
        int height = getMeasuredHeight();
        int w = width;
        int h = height;

        String[] resolution = preferences.getString("displayResolutionExact", "1280x720").split("x");
        w = Integer.parseInt(resolution[0]);
        h = Integer.parseInt(resolution[1]);

        if ((width < height && w > h) || (width > height && w < h))
            p.set(h, w);
        else
            p.set(w, h);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
        
        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(getContext());
        if (preferences.getBoolean("displayStretch", false)) {
            getHolder().setSizeFromLayout();
            return;
        }

        getDimensionsFromSettings();

        if (p.x <= 0 || p.y <= 0)
            return;

        int width = getMeasuredWidth();
        int height = getMeasuredHeight();

        if ((width < height && p.x > p.y) || (width > height && p.x < p.y))
            //noinspection SuspiciousNameCombination
            p.set(p.y, p.x);

        if (width > height * p.x / p.y)
            width = height * p.x / p.y;
        else
            height = width * p.y / p.x;

        getHolder().setFixedSize(p.x, p.y);
        setMeasuredDimension(width, height);
    }
    
    
    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        // TODO: Implement this method
        ramCounter(canvas);
    }
    
    public void ramCounter(Canvas c) {
    Paint paint = new Paint();
        paint.setColor(Color.WHITE);
        paint.setTextSize(15);
        c.drawText("RAM: " + (totalMemory - freeMemory) / BYTES_FOR_MEGABYTES + "/" + (totalMemory / BYTES_FOR_MEGABYTES), 10, 40, paint);
    }
    
    
   private void updateRamCounter() {
    totalMemory = SystemMemoryInfo.getTotalRAM(getContext());
    freeMemory = SystemMemoryInfo.getFreeRAM(getContext());
    invalidate();
    handler.postDelayed(this::updateRamCounter, 50);
}
    

    @Override
    public void sendMouseWheelEvent(float deltaX, float deltaY) {
        sendMouseEvent(deltaX, deltaY, BUTTON_SCROLL, false, true);
    }

    @Override
    public boolean dispatchKeyEventPreIme(KeyEvent event) {
        Activity a = getActivity();
        return (a instanceof EmulationActivity) && ((EmulationActivity) a).handleKey(event);
    }

    // It is used in native code
    void setClipboardText(String text) {
        ClipboardManager clipboard = (ClipboardManager) getContext().getSystemService(Context.CLIPBOARD_SERVICE);
        clipboard.setPrimaryClip(ClipData.newPlainText("X11 clipboard", text));
    }

    static native void connect(int fd);
    native void handleXEvents();
    static native void startLogcat(int fd);
    static native void setClipboardSyncEnabled(boolean enabled);
    static native void sendWindowChange(int width, int height, int framerate);
    public native void sendMouseEvent(float x, float y, int whichButton, boolean buttonDown, boolean relative);
    public native void sendTouchEvent(int action, int id, int x, int y);
    public native boolean sendKeyEvent(int scanCode, int keyCode, boolean keyDown);
    public native void sendTextEvent(byte[] text);
    public native void sendUnicodeEvent(int code);

    static {
        System.loadLibrary("Xlorie");
    }
}

File: app/src/main/java/com/micewine/emu/LorieView.kt
package com.micewine.emu

import android.annotation.SuppressLint
import android.app.Activity
import android.content.ClipData
import android.content.ClipDescription
import android.content.ClipboardManager
import android.content.Context
import android.content.ContextWrapper
import android.graphics.Color
import android.graphics.Point
import android.graphics.drawable.ColorDrawable
import android.text.InputType
import android.util.AttributeSet
import android.util.Log
import android.view.KeyEvent
import android.view.Surface
import android.view.SurfaceHolder
import android.view.SurfaceView
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputConnection
import androidx.annotation.Keep
import com.micewine.emu.activities.EmulationActivity
import com.micewine.emu.activities.MainActivity.Companion.preferences
import com.micewine.emu.activities.MainActivity.Companion.selectedResolution
import com.micewine.emu.input.InputStub
import com.micewine.emu.input.TouchInputHandler
import dalvik.annotation.optimization.CriticalNative
import dalvik.annotation.optimization.FastNative
import java.nio.charset.StandardCharsets

@Keep
@SuppressLint("WrongConstant")
@Suppress("deprecation")
class LorieView : SurfaceView, InputStub {
    interface Callback {
        fun changed(
            sfc: Surface?,
            surfaceWidth: Int,
            surfaceHeight: Int,
            screenWidth: Int,
            screenHeight: Int
        )
    }

    internal interface PixelFormat {
        companion object {
            const val BGRA_8888: Int = 5 // Stands for HAL_PIXEL_FORMAT_BGRA_8888
        }
    }

    private var clipboard: ClipboardManager? = null
    private var lastClipboardTimestamp = System.currentTimeMillis()
    private var mCallback: Callback? = null
    private val p = Point()
    private val mSurfaceCallback: SurfaceHolder.Callback = object : SurfaceHolder.Callback {
        override fun surfaceCreated(holder: SurfaceHolder) {
            holder.setFormat(PixelFormat.BGRA_8888)
        }

        override fun surfaceChanged(holder: SurfaceHolder, f: Int, width: Int, height: Int) {
            Log.d("SurfaceChangedListener", "Surface was changed: " + measuredWidth + "x" + measuredHeight)
            if (mCallback == null) return

            this@LorieView.updateDimensionsFromSettings()
            if (mCallback != null) mCallback!!.changed(holder.surface, measuredWidth, measuredHeight, p.x, p.y)
            this@LorieView.surfaceChanged(holder.surface)
        }

        override fun surfaceDestroyed(holder: SurfaceHolder) {
            if (mCallback != null) mCallback!!.changed(holder.surface, 0, 0, 0, 0)
            this@LorieView.surfaceChanged(holder.surface)
        }
    }

    constructor(context: Context?) : super(context) {
        init()
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
        init()
    }

    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        init()
    }

    @Suppress("unused")
    constructor(
        context: Context?,
        attrs: AttributeSet?,
        defStyleAttr: Int,
        defStyleRes: Int
    ) : super(context, attrs, defStyleAttr, defStyleRes) {
        init()
    }

    private fun init() {
        holder.addCallback(mSurfaceCallback)
        clipboard = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        nativeInit()
    }

    fun setCallback(callback: Callback?) {
        mCallback = callback
        triggerCallback()
    }

    fun regenerate() {
        val callback = mCallback
        mCallback = null
        holder.setFormat(android.graphics.PixelFormat.RGBA_8888)
        mCallback = callback

        triggerCallback()
    }

    fun triggerCallback() {
        isFocusable = true
        isFocusableInTouchMode = true
        requestFocus()

        background =
            object : ColorDrawable(Color.TRANSPARENT) {
                override fun isStateful(): Boolean {
                    return true
                }

                override fun hasFocusStateSpecified(): Boolean {
                    return true
                }
            }

        val r = holder.surfaceFrame
        activity.runOnUiThread {
            mSurfaceCallback.surfaceChanged(
                holder,
                PixelFormat.BGRA_8888,
                r.width(),
                r.height()
            )
        }
    }

    private val activity: Activity
        get() {
            var context = context
            while (context is ContextWrapper) {
                if (context is Activity) {
                    return context
                }
                context = context.baseContext
            }

            throw NullPointerException()
        }

    private fun updateDimensionsFromSettings() {
        val width = measuredWidth
        val height = measuredHeight
        val resolution = (selectedResolution ?: "1280x720" ).split("x".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val w = resolution[0].toInt()
        val h = resolution[1].toInt()

        if ((width < height && w > h) || (width > height && w < h)) {
            p[h] = w
        } else {
            p[w] = h
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)

        if (preferences?.getBoolean("displayStretch", false) == true) {
            holder.setSizeFromLayout()
            return
        }

        updateDimensionsFromSettings()

        if (p.x <= 0 || p.y <= 0) return

        var width = measuredWidth
        var height = measuredHeight

        if ((width < height && p.x > p.y) || (width > height && p.x < p.y))
            p[p.y] = p.x

        if (width > height * p.x / p.y) width = height * p.x / p.y
        else height = width * p.y / p.x

        holder.setFixedSize(p.x, p.y)
        setMeasuredDimension(width, height)

        // In the case if old fixed surface size equals new fixed surface size windowChanged will not be called.
        // We should force it.
        regenerate()
    }

    override fun sendMouseWheelEvent(deltaX: Float, deltaY: Float) {
        sendMouseEvent(deltaX, deltaY, InputStub.BUTTON_SCROLL, false, true)
    }

    override fun dispatchKeyEventPreIme(event: KeyEvent): Boolean {
        if (hardwareKbdScancodesWorkaround) return false
        val a = activity
        return (a is EmulationActivity) && a.handleKey(event)
    }

    private var clipboardListener: ClipboardManager.OnPrimaryClipChangedListener = ClipboardManager.OnPrimaryClipChangedListener { this.handleClipboardChange() }

    fun reloadPreferences() {
        hardwareKbdScancodesWorkaround = false
        clipboardSyncEnabled = true
        setClipboardSyncEnabled(true, true)
        TouchInputHandler.refreshInputDevices()
    }

    // It is used in native code
    fun setClipboardText(text: String?) {
        clipboard!!.setPrimaryClip(ClipData.newPlainText("X11 clipboard", text))

        // Android does not send PrimaryClipChanged event to the window which posted event
        // But in the case we are owning focus and clipboard is unchanged it will be replaced by the same value on X server side.
        // Not cool in the case if user installed some clipboard manager, clipboard content will be doubled.
        lastClipboardTimestamp = System.currentTimeMillis() + 150
    }

    /** @noinspection unused
     */
    // It is used in native code
    fun requestClipboard() {
        if (!clipboardSyncEnabled) {
            sendClipboardEvent("".toByteArray(StandardCharsets.UTF_8))
            return
        }

        val clip = clipboard!!.text
        if (clip != null) {
            val text = clipboard!!.text.toString()
            sendClipboardEvent(text.toByteArray(StandardCharsets.UTF_8))
            Log.d("CLIP", "sending clipboard contents: $text")
        }
    }

    private fun handleClipboardChange() {
        checkForClipboardChange()
    }

    private fun checkForClipboardChange() {
        val desc = clipboard!!.primaryClipDescription
        if (clipboardSyncEnabled && desc != null && lastClipboardTimestamp < desc.timestamp && desc.mimeTypeCount == 1 &&
            (desc.hasMimeType(ClipDescription.MIMETYPE_TEXT_PLAIN) ||
                    desc.hasMimeType(ClipDescription.MIMETYPE_TEXT_HTML))
        ) {
            lastClipboardTimestamp = desc.timestamp
            sendClipboardAnnounce()
            Log.d("CLIP", "sending clipboard announce")
        }
    }

    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) regenerate()

        requestFocus()

        if (clipboardSyncEnabled && hasFocus) {
            clipboard!!.addPrimaryClipChangedListener(clipboardListener)
            checkForClipboardChange()
        } else clipboard!!.removePrimaryClipChangedListener(clipboardListener)

        TouchInputHandler.refreshInputDevices()
    }

    override fun onCreateInputConnection(outAttrs: EditorInfo): InputConnection? {
        outAttrs.inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_FLAG_NO_SUGGESTIONS or InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD

        // Note that IME_ACTION_NONE cannot be used as that makes it impossible to input newlines using the on-screen
        // keyboard on Android TV (see https://github.com/termux/termux-app/issues/221).
        outAttrs.imeOptions = EditorInfo.IME_FLAG_NO_FULLSCREEN

        return super.onCreateInputConnection(outAttrs)
    }

    @FastNative
    private external fun nativeInit()

    @FastNative
    private external fun surfaceChanged(surface: Surface)

    @FastNative
    external fun sendClipboardAnnounce()

    @FastNative
    external fun sendClipboardEvent(text: ByteArray?)

    @FastNative
    external override fun sendMouseEvent(
        x: Float,
        y: Float,
        whichButton: Int,
        buttonDown: Boolean,
        relative: Boolean
    )

    @FastNative
    external override fun sendTouchEvent(action: Int, id: Int, x: Int, y: Int)

    @FastNative
    external override fun sendStylusEvent(
        x: Float,
        y: Float,
        pressure: Int,
        tiltX: Int,
        tiltY: Int,
        orientation: Int,
        buttons: Int,
        eraser: Boolean,
        mouseMode: Boolean
    )

    @FastNative
    external override fun sendKeyEvent(scanCode: Int, keyCode: Int, keyDown: Boolean): Boolean

    @FastNative
    external override fun sendTextEvent(utf8Bytes: ByteArray?)

    companion object {
        private var clipboardSyncEnabled = false
        private var hardwareKbdScancodesWorkaround = false

        @JvmStatic
        external fun renderingInActivity(): Boolean

        @JvmStatic @FastNative
        external fun connect(fd: Int)

        @JvmStatic @CriticalNative
        external fun connected(): Boolean

        @JvmStatic @FastNative
        external fun startLogcat(fd: Int)

        @JvmStatic @FastNative
        external fun setClipboardSyncEnabled(enabled: Boolean, ignored: Boolean)

        @JvmStatic @FastNative
        external fun sendWindowChange(width: Int, height: Int, framerate: Int, name: String?)

        @JvmStatic @FastNative
        external fun requestStylusEnabled(enabled: Boolean)

        @JvmStatic @CriticalNative
        external fun requestConnection()

        init {
            System.loadLibrary("Xlorie")
        }
    }
}
